src/Aspire.Hosting/Devcontainers/DevcontainerPortForwardingLifecycleHook.cs

```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Devcontainers.Codespaces;
+using Aspire.Hosting.Lifecycle;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Aspire.Hosting.Devcontainers;
+
+internal sealed class DevcontainerPortForwardingLifecycleHook : IDistributedApplicationLifecycleHook
+{
+    private readonly ILogger _hostingLogger;
+    private readonly IOptions<CodespacesOptions> _codespacesOptions;
+    private readonly IOptions<DevcontainersOptions> _devcontainersOptions;
+
+    public DevcontainerPortForwardingLifecycleHook(ILoggerFactory loggerFactory, IOptions<CodespacesOptions> codespacesOptions, IOptions<DevcontainersOptions> devcontainersOptions)
+    {
+        _hostingLogger = loggerFactory.CreateLogger("Aspire.Hosting");
+        _codespacesOptions = codespacesOptions;
+        _devcontainersOptions = devcontainersOptions;
+    }
+
+    public async Task AfterEndpointsAllocatedAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken)
```

Comment:
Why do we want to use this event instead of BeforeResourceStart? That way it's not blocking startup of the dashboard.

Reply:
Also even though this executes before the dashboard starts its a local I/O operation so not a huge impact to overall start performance. In my experimentation the dashboard is up and running before I've decided to interact with it.